Add FoT names to CoordMaps

knelli2 opened the pull request:
## Proposed changes

The CoordinateMap(Base) class is now aware of what functions of time are being used for that map. This information can then be used by elements to make decisions based on which maps they have.

Fixes #5247.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Add function_of_time_names to CoordMap base

knelli2 committed:
Add func of time names to time dep maps

knelli2 committed:
Handle func of time name ASSERTs in CoordMap

knelli2 committed:
Appease clang-tidy

wthrowe commented on src/Domain/CoordinateMaps/Composition.cpp:
```diff
@@ -350,7 +358,11 @@ template <typename Frames, size_t Dim, size_t... Is>
 bool Composition<Frames, Dim, std::index_sequence<Is...>>::is_equal_to(
     const CoordinateMapBase<SourceFrame, TargetFrame, Dim>& other) const {
   const auto& cast_of_other = dynamic_cast<const Composition&>(other);
-  bool result = true;
+  bool result =
+      function_of_time_names_ == cast_of_other.function_of_time_names_;
+  if (not result) {
+    return false;
+  }
   EXPAND_PACK_LEFT_TO_RIGHT(
       (result = result and (*get<Is>(maps_) == *get<Is>(cast_of_other.maps_))));
   return result;
```
[optional] Not new in this PR, but this is set of statements is `return ((function_of_time_names_ == cast_of_other.function_of_time_names_) and ... and (*get<Is>(maps_) == *get<Is>(cast_of_other.maps_)));`.

wthrowe commented on src/Domain/CoordinateMaps/CoordinateMap.tpp:
```diff
@@ -25,25 +26,72 @@
+  const auto add_names = [&function_of_time_names, &maps](auto index) {
+    const auto& map = std::get<decltype(index)::value>(maps);
+    using TupleMap = std::decay_t<decltype(map)>;
+    constexpr bool map_is_unique_ptr = tt::is_a_v<std::unique_ptr, TupleMap>;
```
Why are these sometimes unique_ptrs and sometimes not?  What are they when they're not?

knelli2 replied:
The `Composition` coordinate map takes the maps as `unique_ptr<CoordinateMapBase>` while the actual `CoordinateMap` just takes maps as the individual maps (Affine, Rotation, etc...). That's why I needed to handle both here.

wthrowe reviewed (changes requested):
In the future, please separate large reformatting with no content change into separate commits.

knelli2 committed:
clang-format CoordinateMap and test

knelli2 commented:
@wthrowe Added a new commit with `clang-format` stuff (sorry since that file had a lot of changes already I didn't realize I changed a lot of stuff that I didn't touch). Also squashed in your suggestion about the template pack.

knelli2 commented:
@wthrowe Found a mistake in the `pup` function of `ProductOf2Maps` so added a quick fixup that handles it.

wthrowe commented:
That looks right.  Go ahead and squash.

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Shape.cpp:
```diff
@@ -398,6 +376,15 @@ void Shape::pup(PUP::er& p) {
+  if (version >= 1) {
+    p | f_of_t_names_;
+  } else if (p.isUnpacking()) {
+    f_of_t_names_ = std::unordered_set<std::string>{{shape_f_of_t_name_}};
```
Does this pattern of serializing the names work correctly? I think some names are still hardcoded, so if they're changed in the code and the old name is deserialized then there's an inconsistency between the hardcoded name and the name in the set. This means either you don't serialize the names at all but just return the hardcoded names (no need for the member variable even) and don't bother with serialization. Then the serialization will still be broken if names are changed because the domain/maps are inconsistent with the serialized FoT map in the file, but that's also the case on develop right now I think. Or you fix all this by serializing the hardcoded FoT name in the maps. Does that make sense or am I misunderstanding something?

knelli2 replied:
Conclusions of discussion with @nilsvu offline:

The "hardcoded" names are actually in the control system. The control system expects certain names in the functions of time (and the domain creators then must supply those names to be usable with control system). Meaning if we change the control system names (hardcoded in `src/ControlSystems`), then everything breaks. (so let's not do that)

The maps themselves actually take names in their constructors and already serialize them. So what's important here is that when we deserialize the maps and the functions of time, everything is consistent. So since we already serialize the names in the maps, we can just always reconstruct the `std::unordered_set` that is needed now from the serialized names and we don't need to change the versions of the pup functions. So I'll make a change to not serialize the `std::unordered_set` and only reconstruct it.

knelli2 committed:
Simplify is_equal_to in Composition

knelli2 commented:
@nilsvu @wthrowe Made those changes in a fixup

wthrowe commented on src/Domain/CoordinateMaps/TimeDependent/Rotation.cpp:
```diff
@@ -174,6 +150,11 @@ void Rotation<Dim>::pup(PUP::er& p) {
+
+  // No need to pup this because it is uniquely determined by f_of_t_name_
+  if (p.isUnpacking()) {
+    f_of_t_names_.insert(f_of_t_name_);
```
I don't know whether we ever unpack into something that hasn't just been default-constructed, but probably best to clear this first just in case.

wthrowe commented on src/Domain/CoordinateMaps/TimeDependent/Shape.cpp:
```diff
@@ -395,8 +372,13 @@ void Shape::pup(PUP::er& p) {
+  // No need to pup these because they are uniquely determined by other members
   if (p.isUnpacking()) {
     ylm_ = ylm::Spherepack(l_max_, m_max_);
+    f_of_t_names_.insert(shape_f_of_t_name_);
```
Again, clear first.

wthrowe commented on src/Domain/CoordinateMaps/TimeDependent/SphericalCompression.cpp:
```diff
@@ -296,6 +290,11 @@ void SphericalCompression<InteriorMap>::pup(PUP::er& p) {
+
+  // No need to pup this because it is uniquely determined by f_of_t_name_
+  if (p.isUnpacking()) {
+    f_of_t_names_.insert(f_of_t_name_);
```
Again, clear.

wthrowe commented on src/Domain/CoordinateMaps/TimeDependent/Translation.cpp:
```diff
@@ -286,6 +280,11 @@ void Translation<Dim>::pup(PUP::er& p) {
+
+  // No need to pup this because it is uniquely determined by f_of_t_name_
+  if (p.isUnpacking()) {
+    f_of_t_names_.insert(f_of_t_name_);
```
Clear again.

wthrowe reviewed (changes requested):
Squash.

knelli2 commented:
Ok squashed

wthrowe commented:
Clang-tidy failures are boost.
